Maven plugin prints (but does NOT fail due to) a ClassNotFoundException: org.apache.commons.io.file.PathUtils

_(This message has been edited from an originally wrong interpretion of cause and effect.)_

The `mariaDB4j-maven-plugin` prints the following: 

```
    [INFO] [INFO] --- mariaDB4j-maven-plugin:2.5.0:stop (post-integration-test) @ start-stop ---
    [INFO] [INFO] Stopping MariaDB4j...
    [INFO] [INFO] Program [/tmp/MariaDB4j/base/bin/mysqld, --no-defaults, --console, --skip-grant-tables, --max_allowed_packet=64M, --basedir=/tmp/MariaDB4j/base, --datadir=/tmp/MariaDB4j/data/42757, --port=42757, --socket=/tmp/MariaDB4j.42757.sock] (in working directory /tmp/MariaDB4j/base) just exited, with value 0
    [INFO] [INFO] Successfully destroyed Program [/tmp/MariaDB4j/base/bin/mysqld, --no-defaults, --console, --skip-grant-tables, --max_allowed_packet=64M, --basedir=/tmp/MariaDB4j/base, --datadir=/tmp/MariaDB4j/data/42757, --port=42757, --socket=/tmp/MariaDB4j.42757.sock] (in working directory /tmp/MariaDB4j/base)
    [INFO] [INFO] Database stopped.
    [INFO] [INFO] ------------------------------------------------------------------------
    [INFO] [INFO] BUILD SUCCESS
    [INFO] [INFO] ------------------------------------------------------------------------
    [INFO] [INFO] Total time:  5.672 s
    [INFO] [INFO] Finished at: 2021-11-28T22:21:48+01:00
    [INFO] [INFO] ------------------------------------------------------------------------
    [INFO] [INFO] cleanupOnExit() ShutdownHook quietly deleting temporary DB data directory: /tmp/MariaDB4j/data/42757
    [INFO] Exception in thread "Shutdown Hook Deletion Thread for Temporary DB /tmp/MariaDB4j/data/42757" java.lang.NoClassDefFoundError: org/apache/commons/io/file/PathUtils
    [INFO]      at org.apache.commons.io.FileUtils.forceDelete(FileUtils.java:1341)
    [INFO]      at org.apache.commons.io.FileUtils.cleanDirectory(FileUtils.java:324)
    [INFO]      at org.apache.commons.io.FileUtils.deleteQuietly(FileUtils.java:1236)
    [INFO]      at ch.vorburger.mariadb4j.DB$1.run(DB.java:474)
    [INFO] Caused by: java.lang.ClassNotFoundException: org.apache.commons.io.file.PathUtils
    [INFO]      at org.codehaus.plexus.classworlds.strategy.SelfFirstStrategy.loadClass(SelfFirstStrategy.java:50)
    [INFO]      at org.codehaus.plexus.classworlds.realm.ClassRealm.unsynchronizedLoadClass(ClassRealm.java:271)
    [INFO]      at org.codehaus.plexus.classworlds.realm.ClassRealm.loadClass(ClassRealm.java:247)
    [INFO]      at org.codehaus.plexus.classworlds.realm.ClassRealm.loadClass(ClassRealm.java:239)
    [INFO]      ... 4 more
```

~~This should be made to not appear only on `mvn release:prepare`, but for every `mvn package`, so that whatever PR broke this (possibly #477 or #478 or #482 or #483 today) would fail to pass CI for this to be noticed much earlier.~~

@robin-xyzt-ai given your interest in the Maven Plugin in #476, would you be willing to contrib. a PR for this?

~~This blocks #334.~~

To my surprise, creating my own `FileUtils` class which contains all the necessary code just results in

```
[INFO] Exception in thread "Shutdown Hook Deletion Thread for Temporary DB /tmp/MariaDB4j/data/46521" java.lang.NoClassDefFoundError: ch/vorburger/mariadb4j/FileUtils
[INFO]  at ch.vorburger.mariadb4j.DB$1.run(DB.java:474)
[INFO] Caused by: java.lang.ClassNotFoundException: ch.vorburger.mariadb4j.FileUtils
[INFO]  at org.codehaus.plexus.classworlds.strategy.SelfFirstStrategy.loadClass(SelfFirstStrategy.java:50)
[INFO]  at org.codehaus.plexus.classworlds.realm.ClassRealm.unsynchronizedLoadClass(ClassRealm.java:271)
[INFO]  at org.codehaus.plexus.classworlds.realm.ClassRealm.loadClass(ClassRealm.java:247)
[INFO]  at org.codehaus.plexus.classworlds.realm.ClassRealm.loadClass(ClassRealm.java:239)
```

There clearly isn't much what you can do in a shut down hook. Perhaps it works if I put all the code in a single file,

--- a/mariaDB4j-core/src/main/java/ch/vorburger/mariadb4j/DB.java
+++ b/mariaDB4j-core/src/main/java/ch/vorburger/mariadb4j/DB.java
@@ -451,35 +451,7 @@ public class DB {
     protected void cleanupOnExit() {
         String threadName = "Shutdown Hook Deletion Thread for Temporary DB " + dataDir.toString();
         final DB db = this;
-        Runtime.getRuntime().addShutdownHook(new Thread(threadName) {
-
-            @Override
-            public void run() {
-                // ManagedProcess DestroyOnShutdown ProcessDestroyer does
-                // something similar, but it shouldn't hurt to better be save
-                // than sorry and do it again ourselves here as well.
-                try {
-                    // Shut up and don't log if it was already stop() before
-                    if (mysqldProcess != null && mysqldProcess.isAlive()) {
-                        logger.info("cleanupOnExit() ShutdownHook now stopping database");
-                        db.stop();
-                    }
-                } catch (ManagedProcessException e) {
-                    logger.warn("cleanupOnExit() ShutdownHook: An error occurred while stopping the database", e);
-                }
-
-                if (dataDir.exists() && (configuration.isDeletingTemporaryBaseAndDataDirsOnShutdown()
-                                    && Util.isTemporaryDirectory(dataDir.getAbsolutePath()))) {
-                    logger.info("cleanupOnExit() ShutdownHook quietly deleting temporary DB data directory: " + dataDir);
-                    FileUtils.deleteQuietly(dataDir);
-                }
-                if (baseDir.exists() && (configuration.isDeletingTemporaryBaseAndDataDirsOnShutdown()
-                                    && Util.isTemporaryDirectory(baseDir.getAbsolutePath()))) {
-                    logger.info("cleanupOnExit() ShutdownHook quietly deleting temporary DB base directory: " + baseDir);
-                    FileUtils.deleteQuietly(baseDir);
-                }
-            }
-        });
+        Runtime.getRuntime().addShutdownHook(new DBShutdownHook(threadName, db, () -> mysqldProcess, () -> baseDir, () -> dataDir, configuration));
     }
 
     // The dump*() methods are intentionally *NOT* made "synchronized",
@@ -533,4 +505,5 @@ public class DB {
         builder.setDestroyOnShutdown(true);
         return builder.build();
     }
+
 }

--- a/mariaDB4j-core/src/main/java/ch/vorburger/mariadb4j/DBShutdownHook.java
+++ b/mariaDB4j-core/src/main/java/ch/vorburger/mariadb4j/DBShutdownHook.java
@@ -1,0 +1,303 @@
+package ch.vorburger.mariadb4j;
+/*
+ * #%L
+ * MariaDB4j
+ * %%
+ * Copyright (C) 2012 - 2021 Michael Vorburger
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+import ch.vorburger.exec.ManagedProcess;
+import ch.vorburger.exec.ManagedProcessException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.FileFilter;
+import java.io.IOException;
+import java.nio.file.*;
+import java.nio.file.attribute.*;
+import java.util.*;
+import java.util.function.Supplier;
+
+/**
+ * During shutdown, the classloader doesn't provide access to the majority of the classes.
+ * As such, we put all the code required for the shutdown in a single class and don't use any external library code.
+ * <p>
+ * For the same reason this class implements {@code FileVisitor}.
+ * Normally we would use an additional class for that (anonymous or (static) inner), but the classloader doesn't find that class.
+ * <p>
+ * Everything outside the {@link #run()} method is a partial copy from {@link org.apache.commons.io.FileUtils} implementation
+ * to delete files.
+ * <p>
+ * See https://github.com/vorburger/MariaDB4j/issues/488.
+ */
+class DBShutdownHook extends Thread implements FileVisitor<Path>{
+
+    private static final Logger logger = LoggerFactory.getLogger(DB.class);
+
+    private final DB db;
+    private final Supplier<ManagedProcess> mysqldProcessSupplier;
+    private final Supplier<File> dataDirSupplier;
+    private final Supplier<File> baseDirSupplier;
+    private final DBConfiguration configuration;
+    private final LinkOption[] linkOptions = new LinkOption[0];
+    
+    public DBShutdownHook(String threadName, DB db, Supplier<ManagedProcess> mysqldProcessSupplier, Supplier<File> baseDirSupplier, Supplier<File> dataDirSupplier, DBConfiguration configuration) {
+        super(threadName);
+        this.db = db;
+        this.mysqldProcessSupplier = mysqldProcessSupplier;
+        this.baseDirSupplier = baseDirSupplier;
+        this.dataDirSupplier = dataDirSupplier;
+        this.configuration = configuration;
+    }
+
+    private boolean deleteQuietly(File file) {
+        if (file == null) {
+            return false;
+        } else {
+            try {
+                if (file.isDirectory()) {
+                    cleanDirectory(file);
+                }
+            } catch (Exception ignore) {
+            }
+
+            try {
+                return file.delete();
+            } catch (Exception ignore) {
+                return false;
+            }
+        }
+    }
+
+    private void cleanDirectory(File directory) throws IOException {
+        File[] files = listFiles(directory, null);
+        List<Exception> causeList = new ArrayList<>();
+        for (File file : files) {
+            try {
+                forceDelete(file);
+            } catch (IOException exception) {
+                causeList.add(exception);
+            }
+        }
+
+        if (!causeList.isEmpty()) {
+            IOException ioException = new IOException(directory.toString());
+            causeList.forEach(ioException::addSuppressed);
+            throw ioException;
+        }
+    }
+
+    private void forceDelete(File file) throws IOException {
+        Objects.requireNonNull(file, "file");
+
+        try {
+            delete(file.toPath());
+        } catch (IOException exception) {
+            throw new IOException("Cannot delete file: " + file, exception);
+        }
+    }
+
+    private void delete(Path path) throws IOException {
+        if (Files.isDirectory(path, linkOptions)) {
+            deleteDirectory(path);
+        } else {
+            deleteFile(path);
+        }
+    }
+
+    private void deleteDirectory(Path directory) throws IOException {
+        Files.walkFileTree(directory, this);
+    }
+
+    private void deleteFile(Path file) throws IOException {
+        if (Files.isDirectory(file, linkOptions)) {
+            throw new NoSuchFileException(file.toString());
+        } else {
+            boolean exists = Files.exists(file, linkOptions);
+            if (exists) {
+                setReadOnly(file, false);
+            }
+
+            Files.deleteIfExists(file);
+        }
+    }
+
+    private Path setReadOnly(Path path, boolean readOnly) throws IOException {
+        List<Exception> causeList = new ArrayList<>(2);
+        DosFileAttributeView fileAttributeView = Files.getFileAttributeView(path, DosFileAttributeView.class, linkOptions);
+        if (fileAttributeView != null) {
+            try {
+                fileAttributeView.setReadOnly(readOnly);
+                return path;
+            } catch (IOException exception) {
+                causeList.add(exception);
+            }
+        }
+
+        PosixFileAttributeView posixFileAttributeView = Files.getFileAttributeView(path, PosixFileAttributeView.class, linkOptions);
+        if (posixFileAttributeView != null) {
+            PosixFileAttributes readAttributes = posixFileAttributeView.readAttributes();
+            Set<PosixFilePermission> permissions = readAttributes.permissions();
+            permissions.remove(PosixFilePermission.OWNER_WRITE);
+            permissions.remove(PosixFilePermission.GROUP_WRITE);
+            permissions.remove(PosixFilePermission.OTHERS_WRITE);
+
+            try {
+                return Files.setPosixFilePermissions(path, permissions);
+            } catch (IOException exception) {
+                causeList.add(exception);
+            }
+        }
+
+        if (!causeList.isEmpty()) {
+            IOException ioException = new IOException(path.toString());
+            causeList.forEach(ioException::addSuppressed);
+            throw ioException;
+        } else {
+            throw new IOException(String.format("No DosFileAttributeView or PosixFileAttributeView for '%s' (linkOptions=%s)", path, Arrays.toString(linkOptions)));
+        }
+    }
+
+    private File[] listFiles(File directory, FileFilter fileFilter) throws IOException {
+        requireDirectoryExists(directory, "directory");
+        File[] files = fileFilter == null ? directory.listFiles() : directory.listFiles(fileFilter);
+        if (files == null) {
+            throw new IOException("Unknown I/O error listing contents of directory: " + directory);
+        } else {
+            return files;
+        }
+    }
+
+    private File requireDirectoryExists(File directory, String name) {
+        requireExists(directory, name);
+        requireDirectory(directory, name);
+        return directory;
+    }
+
+    private File requireExists(File file, String fileParamName) {
+        Objects.requireNonNull(file, fileParamName);
+        if (!file.exists()) {
+            throw new IllegalArgumentException("File system element for parameter '" + fileParamName + "' does not exist: '" + file + "'");
+        } else {
+            return file;
+        }
+    }
+
+    private File requireDirectory(File directory, String name) {
+        Objects.requireNonNull(directory, name);
+        if (!directory.isDirectory()) {
+            throw new IllegalArgumentException("Parameter '" + name + "' is not a directory: '" + directory + "'");
+        } else {
+            return directory;
+        }
+    }
+
+    private boolean isEmptyDirectory(Path directory) throws IOException {
+        DirectoryStream<Path> directoryStream = Files.newDirectoryStream(directory);
+        Throwable throwable = null;
+
+        boolean hasNext;
+        try {
+            hasNext = !directoryStream.iterator().hasNext();
+        } catch (Throwable exception) {
+            throwable = exception;
+            throw exception;
+        } finally {
+            if (directoryStream != null) {
+                if (throwable != null) {
+                    try {
+                        directoryStream.close();
+                    } catch (Throwable exception) {
+                        throwable.addSuppressed(exception);
+                    }
+                } else {
+                    directoryStream.close();
+                }
+            }
+
+        }
+
+        return hasNext;
+    }
+
+    @Override
+    public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException {
+        return FileVisitResult.CONTINUE;
+    }
+
+    @Override
+    public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+        if (Files.exists(file, linkOptions)) {
+            setReadOnly(file, false);
+            Files.deleteIfExists(file);
+        }
+
+        if (Files.isSymbolicLink(file)) {
+            try {
+                Files.delete(file);
+            } catch (NoSuchFileException ignore) {
+            }
+        }
+        return FileVisitResult.CONTINUE;
+    }
+
+    @Override
+    public FileVisitResult visitFileFailed(Path file, IOException exc) throws IOException {
+        Objects.requireNonNull(file);
+        throw exc;
+    }
+
+    @Override
+    public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOException {
+        if (isEmptyDirectory(dir)) {
+            Files.deleteIfExists(dir);
+        }
+        Objects.requireNonNull(dir);
+        if (exc != null)
+            throw exc;
+        return FileVisitResult.CONTINUE;
+    }
+
+    @Override
+    public void run() {
+        ManagedProcess mysqldProcess = mysqldProcessSupplier.get();
+        File dataDir = dataDirSupplier.get();
+        File baseDir = baseDirSupplier.get();
+        // ManagedProcess DestroyOnShutdown ProcessDestroyer does
+        // something similar, but it shouldn't hurt to better be save
+        // than sorry and do it again ourselves here as well.
+        try {
+            // Shut up and don't log if it was already stop() before
+            if (mysqldProcess != null && mysqldProcess.isAlive()) {
+                logger.info("cleanupOnExit() ShutdownHook now stopping database");
+                db.stop();
+            }
+        } catch (ManagedProcessException e) {
+            logger.warn("cleanupOnExit() ShutdownHook: An error occurred while stopping the database", e);
+        }
+
+        if (dataDir.exists() && (configuration.isDeletingTemporaryBaseAndDataDirsOnShutdown()
+                && Util.isTemporaryDirectory(dataDir.getAbsolutePath()))) {
+            logger.info("cleanupOnExit() ShutdownHook quietly deleting temporary DB data directory: " + dataDir);
+            deleteQuietly(dataDir);
+        }
+        if (baseDir.exists() && (configuration.isDeletingTemporaryBaseAndDataDirsOnShutdown()
+                && Util.isTemporaryDirectory(baseDir.getAbsolutePath()))) {
+            logger.info("cleanupOnExit() ShutdownHook quietly deleting temporary DB base directory: " + baseDir);
+            deleteQuietly(baseDir);
+        }
+    }
+}

